### PR TITLE
[WIP] Support NHWC tensor layout (2)

### LINKED ISSUE #5076: Support NHWC mode to exploit full potential of TensorCore
When TensorCore is available, the Convolution operator of cuDNN will internally convert the input tensor to NHWC order from NCHW first, then calculate convolution results, and convert back the resulting tensor into NCHW again. Those two converting phases take a lot of time in a general image classification workload like ResNet50 training on ImageNet-1K. It's better to have NHWC mode in Chainer to prevent the conversion before and after the TensorCore calculation.

ref: https://devblogs.nvidia.com/tensor-core-ai-performance-milestones/

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #5076
Requires cupy/cupy#2235

This is yet another design candidate to support NHWC tensor layout for cuDNN (related: #7375)

### CODE REVIEW

**Comment by emcastillo** [comment]
Main differences with #7375 
This PR:
+ Memory layout is explicitly tracked by `Variable` objects 
+ Uses a global config parameter to set `layout` when building links (using a context manager)
+ Require explicit layout conversions on the forward method using `Variable.as_layout`

#7375
+ Layout is only tracked inside the function objects
+ Uses a global config parameter to set `layout` when applying functions
+ Require explicit layout conversions on the forward method using a Transpose derived function

My thoughts
+ #7375 is simpler in terms of code, but interface is convoluted.
+ I think that layout should not be a parameter of the links themselves. I think it is good to have it in the variable.
+ I wonder whether it is possible to get rid of the initial context manager?
A possible solution could be first to transpose the tensor using the as_layout. `Link`s should detect whether their internal state is in the same layout as the input tensor and then transpose it accordingly. Functions should raise an error if all the input variables do not share the same layout.

**Comment by niboshi** [comment]
I'm sorry for late response.

>I wonder whether it is possible to get rid of the initial context manager?

This is needed because links have to initialize their parameters with the correct layout in `__init__`.

**Comment by emcastillo** [comment]
> This is needed because links have to initialize their parameters with the correct layout in __init__.
Can we make links to dynamically change their layout depending on the input tensor layout?

**Comment by niboshi** [comment]
Do you mean transposing weights in `forward` every time? The overhead would be prohibitive.

**Comment by emcastillo** [comment]
No, transpose and let the parameters remain transposed within the link, however as we just discussed there is the problem that the link parameters will change their layout after initialization and this can lead to inconsistent behavior if users try to directly observe the parameters.

**Comment by niboshi** [comment]
Please merge #8462 first.

File: chainer/links/connection/convolution_2d.py [hunk 1]
```diff
@@ -227,8 +238,14 @@ def forward(self, x):
             ~chainer.Variable: Output of the convolution.
 
         """
-        if self.W.array is None:
-            self._initialize_params(x.shape[1])
+        x = chainer.as_variable(x)
+        assert x.layout == self.x_layout
+        # self.W can be a Variable instead of Parameter: #8462
+        # TODO(niboshi): Use Parameter.is_initialized.
+        if self.W.raw_array is None:
```
Comment (niboshi) [inline chainer/links/connection/convolution_2d.py]: I fixed to avoid assuming that `self.W` is a `Parameter`.
#8462 is no longer required.

**Comment by niboshi** [comment]
I think implementation (except for tests and documentation) is done.
Could you take another look? @emcastillo

**Comment by niboshi** [comment]
To benefit from NHWC, you need to set environment variables
* `CHAINER_CUDNN_FAST_BATCH_NORMALIZATION=1`
* `CHAINER_DTYPE=float16`

and run `train_imagenet.py` with
* `--arch resnet50_nhwc`
* and relatively large batch size (e.g. `--batch 256`).

**Comment by emcastillo** [comment]
Sure! thanks a lot for the great effort 🙇

File: chainer/backends/cuda.py [hunk 2]
```diff
@@ -793,3 +793,17 @@ def get_cudnn_dropout_states_core(thread_id):
...
+
+def _get_cudnn_tensor_layout_x(x_layout):
+    if x_layout == chainer.memory_layouts.CUDNN_CHANNEL_FIRST_X:
+        return cuda.cudnn.CUDNN_TENSOR_NCHW
+    assert x_layout == chainer.memory_layouts.CUDNN_CHANNEL_LAST_X
+    return cuda.cudnn.CUDNN_TENSOR_NHWC
+
+
+def _get_cudnn_tensor_layout_w(w_layout):
+    if w_layout == chainer.memory_layouts.CUDNN_CHANNEL_FIRST_W:
```
Comment (emcastillo) [inline chainer/backends/cuda.py]: Is there any specific reason why these two functions are separated?

Comment (niboshi) [reply chainer/backends/cuda.py]: cuDNN APIs accept `d_layout` and `w_layout` separately, and in my intuition it's just a "coincidence" that values of these two sets of arguments are identical in cuDNN.

Comment (emcastillo) [reply chainer/backends/cuda.py]: Ok, It is safer to leave them as it is now then. It is not a guaranteed behavior

**Comment by emcastillo** [comment]
I verified the performance of the implementation
There is no perf degradation wrt to the master branch

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Failures seem to be related

**Comment by niboshi** [comment]
Fixed: 45822b9d1f527d15b744d034b27813368337c6c1 c9cd4cabd61b57dfe23843b655f301f89f0aecca 53af8b960683f8f8b32d73cac3f73f6b3acd3393

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
pfnCI failures seem to be related:
`AttributeError: module 'cupy.cudnn' has no attribute 'batch_normalization_forward_training_ex'`

Seems like is using cupy from pypi instead of master?

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**